### ISSUE TITLE
Writes to observed properties in an Observer's builder fails to rebuild the widget

### ISSUE THREAD

**Issue by shyndman** [mobxjs/mobx.dart#296]
Take the following simplified example:

```dart
return Observer(
  builder: (_) {
    // Marks observableValue as observed by the reaction
    final value = store.observableValue;
    // Adds the reaction as pending to the current batch
    store.someValue = 123;
    return Widget(value);
  },
);
```

What happens next is that the `Observer`'s reaction runs, calling `setState()`, but because the observer is still being actively built, `setState()` is a noop, and `build()` never re-runs, nor does the tracking function. 

And that's it. `Observer` is dead in the water.

**Comment by rrousselGit** [comment]
That's to be expected.
The build method is supposed to be pure. You shouldn't modify variables inside it.

**Comment by shyndman** [comment]
Even if that's the case, we shouldn't be failing silently.

**Comment by pavanpodila** [comment]
I was about to post the same as @rrousselGit. Build methods should have no mutations. In fact we should throw an exception here?

**Comment by shyndman** [comment]
Because that example was simple, but in the case we encountered it, it was very difficult to track down. 

A model object was being requested from a service several layers deep. The value didn't yet exist, an async request went out to fetch it and a placeholder value was returned. When the async value came back, it wrote back to the map. nothing updated. This was deep in our machinery and it wasn't immediately obvious what was happening.

In fact, to the caller of this service, the call was effectively pure. We were in essence setting a default value for something. Yes, it had a side-effect, but it was convenient for our case. You'd have a hard time convincing me that this is wrong actually.

But if it breaks things, then I have to know about it.

**Comment by rrousselGit** [comment]
To begin with, shouldn't mutations outside of an action be disabled by default? And actions being disabled inside Observer?

**Comment by rrousselGit** [comment]
The "silently failing" is the same behavior as what Flutter does.
It's technically not a fail. It doesn't always cause issues.

The only situation where it'll cause issues is wheen the store is coming from a parent widget. 
In which case, yes, we should throw. 
But if it's a local store, then it's definitely fine.

**Comment by shyndman** [comment]
Well in their case it isn't a failure at all. It's just a noop. That's fine.

It's a failure in our case because the MobX Observer's invalidation function fails to track the next iteration, meaning the Observer is marked stale and never becomes up to date. No future change to the store will have any effect, because it's never enqueued as a pending reaction.

**Comment by shyndman** [comment]
Maybe a repro would help?

**Comment by pavanpodila** [comment]
Frankly mutations inside a reaction do wait until the batch completes to trigger any change notifications, which will be caught by the same or other reactions for a re-execution. Ya, a repro would certainly help. 

Technically a mutation inside a reaction can re-trigger the same reaction. In fact, we allow this cyclic dependency but the cycle should bail out before 100 loops.

**Comment by shyndman** [comment]
Yes, and it does retrigger — by calling `setState`. But `setState` in this instance doesn't have the intended effect of calling `build` again.

Sure. I'll put something together tonight.

**Comment by pavanpodila** [comment]
I added a separate issue to do the error reporting for mutations inside a builder: #297

**Comment by shyndman** [comment]
I disappeared, sorry. I'll post a repro tomorrow.

**Comment by pavanpodila** [comment]
Also I think writes can be disabled during the execution of the builder, which can help prevent subtle mutation related bugs like the one you found. This kind of [mutual exclusivity of reads and writes was introduced in Swift 5.0](https://swift.org/blog/swift-5-exclusivity/) to prevent similar kind of memory issues.
